README
Installation
============

This software requires Python 2.6 or 2.7.

# the installer will perform these commands,
# as a human, you will do these yourself,

curl -o /tmp/virtualenv.py 'https://raw.github.com/pypa/virtualenv/master/virtualenv.py'

# for testing 2.6
python2.6 /tmp/virtualenv.py ENV26
. ENV26/bin/activate

# for testing 2.7,
python2.7 /tmp/virtualenv.py ENV27
. ENV27/bin/activate

pip install -r requirements.txt


Getting Started
===============

To start the system, launch 'x84'.

	./x84

Optional command line arguments,
 -cfg <configuration file>
 -v enable debug logging


Connecting
==========

telnet localhost 6023


Compatible Clients (UTF-8)
============================

iTerm,
- Menu item iTerm -> Preferences,
  section Profiles, select 'Text' tab,
  chose 'Andale Mono' font.
 
PuTTy
- Under preference item Window -> Translation,
  option 'Remote character set',
  change 'iso8859-1' to 'UTF-8'.

Terminal.app
 - Menu item Terminal -> Preferences,
   chose profile 'Pro', (Font Andale Mono),
   enable 'use bright colors for bold text'.

uxterm


Semi-compatible Clients (CP437)
=================================

SyncTerm
- Select cp437 when prompted by the bbs system (charset.py).

Other 7-bit clients
- Select cp437 when prompted by the bbs system (charset.py).
- Use a font of cp437 encoding, such as 'Terminus'.


Monitoring
==========

Sessions are recorded to ttyrecordings/ folder, and can be played with
'ttyplay' or compatible utility. The '-p' option can be used to monitor
live sessions, analogous to 'tail -f'.


Session Scripts
===============

The default.ini 'scriptpath' of [session] defines folder 'default/', the default
lookup path for goto() and gosub() commands executed by bbs scripts. This path
should also contain the 'matrix.py' script.

By default, matrix.py will chain to 'nua.py' for new account creation, and
'main.py' for a main menu. The 'topscript' script, 'top.py' should receive
the handle of the user that is authorized, and log the call. If 'topscript'
is called without a handle, the value will default to 'None', and the user
is presumed 'anonymous'.


Matrix
======

The default.ini section 'matrix' describes the 'script' path of a python
program that is loaded when a telnet connection is first made. This script
should initialize a login handshake, or otherwise approve the connection for
continuation.

This section also defines a 'topscript' python program, which is the script to
execute after the 'matrix' routine has determined succesful authentication.


Example scripts
===============

These scripts are found in the default/ subdirectory

  charset.py    prompt user to select 'cp437' or 'utf-8' character set
  lc.py         display 'last callers' of the board
  logoff.py     an automsg display/prompt and logoff
  main.py       main menu, called by default topscript, top.py
  matrix.py     login script, called by new connections
  news.py       file reader used to display 'system news'
  nua.py        new user application routine called by default matrix, matrix.py
  ol.py         bbs-scene.org enabled 'one liners' / wall script
  si.py         display system information door
  speedhack.py  an alternative 'topscript', to simulate dgamelaunch+NAO
  telnet.py     a telnet-out door, pass telnet address/port  as arguments
  top.py        default 'top script' called after authentication success
  weather.py    a weather API and lightbar demonstration .. ?


Globals
=======

bbs helper functions are exported to the the global namespace of all bbs scripts.
these scripts can be found in the 'bbs' sub-folder. the special __init__.py
file defines a special list, '__all__', containing a list of terms to be
injected into all bbs session scripts. it is as if the statement:

from bbs import *

is applied to all bbs scripts executed by goto() and gosub() commands.

An irc channel, '#prsv' on efnet, is available for development discussion.


CHANGES
2.0.16
  - This is the *final* release of x/84, there will be no more updates!!
  - update all setup.py requirements to the latest, here's hoping the
    direct downstream dependencies honor SEMVER and this continues to
    work for many more years to come!
  - removed bbs.modem api and xmodem dependencies, it is not used
  - removed 'weather' from main menu, it is broken
  - removed 'sysop' from main menu, the "x84 message net" is dead
  - added 'ct', a 24-bit color example program (from blessed)
2.0.15
  - bugfix: upgraded six version to 1.9.0 to meet blessed requirements
  - enhancement: added 'raw' flag to x84.bbs.door.Door to send raw output to
    the session.write method. useful for incorporating external file transfer
    programs.
  - refactor: all example scripts now prefer getterminal().inkey() rather than
    deprecated getch(), getch() returns multiple types, None for timeout,
    keycode as integer for application keys, or string for characters.
    inkey() interface always returns a string, sometimes empty, with
    .code attribute for application keys.
  - This refactoring allows us to use the latest version of 'blessed'
    library.
2.0.14
  - enhancement: ability to connect to irc servers with password protection
    using 'password' setting in 'irc' section of default.ini
  - enhancement: pipe codes for colors/modes are now supported in ircchat.py.
  - bugfix: msgarea.py fix unread message count (only include messages the
    user can view)
2.0.13
  - bugfix: previously the 'stop/more/continuous' prompt was not pausing news
    output in the appropriate place. prompt_pager also deals with wrapped lines
    much better now.
  - bugfix: OpenSSL has changed a method signature in their library, which
    broke the versions of cffi and cryptography x84 relies on. increased the
    required versions of these libraries.
2.0.12
  - enhancement: [system] option scriptpath now supports a comma
    separated list of directories in which customizations are stored.
2.0.11
  - bugfix: weather.py fix incorrect postal code return from Accu Weather.
    Changed the default to populate search by city/state instead of
    postal since this is no longer returned from accu weather correctly.
    Also re-added missing save location field that was outside of
    while loop to funtion exit.
  - bugfix: tetris.py fix for sending terminal font sequence prior to
    ansi screen display instead of after.
2.0.10
  - bugfix: display updates when posting to oneliners, issue #235

2.0.9
  - bugfix: profile.py now accepts a target handle as an optional
    argument to its main() function, so online.py can successfully
    launch the profile editor for online users.
2.0.8
  - bugfix: Fixed a nasty Return value bug in vote.py that could
    cause a database failure.
  - bugfix: XML element tag name in weather.py AccuWeather response now
    includes namespace
  - bugfix: Refresh screen error in msgarea.py
  - workaround: always force refresh on event 'gosub' as terminal clients
    such as syncterm, netrunner, and etherterm do not support the
    "alt screen" sequence well, or at all.
  - sometimes (when TERM is 'ansi') sesame doors did not run.
2.0.7
  - bugfix: ValueError in profile.py for setting bad timeout value
  - bugfix: ASCII colly diz with extended chars was causing an error
  - bugfix: KeyError in hackernews.py
  - deprecation: the "extras" subfolder of default script has been
    moved to https://github.com/x84-extras

  (version 2.0.6 was not released to pypi)

2.0.5
  - bugfix: UnicodeDecodeError when quoting some messages
  - bugfix: '__uploads__' folder not hidden by sftp as intended
2.0.4
  - bugfix: sesame did not honor want_cols/rows when
    the given terminal size was already of its exact
    dimensions.
2.0.3
  - bugfix: import common.py as relative from main.py
2.0.1
  - bugfix: did not include default 'top' art
  - bugfix: could not press 'return' to continue in sesame
2.0.0
  - you will need to ensure to set 'enabled = yes' for the shroo-ms api,
    previously this was enabled if the section alone existed
  - you will need to change the values of `newcmds', `anoncmds',
    and 'byecmds' of section [matrix] to be comma-delimited,
    not space-delimited as before. Similarly, you must also change
    value of 'invalid_handles' in section [nua].
  - the meaning of [system] option 'termcap-ansi', when not valued 'no', now
    coerces any reported terminal types *beginning* with 'ansi' to
    the target type.
  - fail2ban is enabled by [fail2ban] option 'enabled = yes'. Previously,
    the mere existence of section [fail2ban] was sufficient.
  - *new* option, 'anoncmds' in section 'matrix', where previously only
    the 'anonymous' user was a legitimate synonym, you may now provide
    many different ones -- humorously, you could provide 'root'.
  - *new* option, 'script_ssh' in section 'matrix', points to a new
    login matrix script that takes new keyword arguments, 'anonymous',
    or 'new' as booleans True or False, or user keyword argument as
    string representing the (already authenticated) user handle.
  - keyword arguments for gosub() and goto() are now supported.
  - ssh key authentication support
  - *removed*: x84.bbs.log.ColoredConsoleHandler.
    If you see exception: ``ImportError: No module named log``
    simply change line, ``class = x84.bbs.log.ColoredConsoleHandler``
    to new line, ``class = logging.StreamHandler``
    in your ~/.x84/logging.ini or /etc/x84/logging.ini file!
  - new bulletins, userlist, from hellbeard
  - new door 'sesame.py' script, from maze
  - project 'blessed' has been merged. Some minor API changes that
    currently just emit a DeprecationWarning, will be removed in 2.0,
    especially in regards to sequence-safe textwrap() function, one
    should instead use term.wrap().
  - project 'wcwidth' has been merged. No change.
  - move_x now works for screen/etc. TERM types (fixed in blessed).

engine.py
#!/usr/bin/env python
"""
command-line launcher and main event loop for x/84
"""
# Please place _ALL_ Metadata here. No need to duplicate this
# in every .py file of the project -- except where individual scripts
# are authored by someone other than the authors, licensing differs, etc.
__author__ = "Johannes Lundberg, Jeffrey Quast"
__copyright__ = "Copyright 2012"
__credits__ = ["Johannes Lundberg", "Jeffrey Quast",
               "Wijnand Modderman-Lenstra", "zipe", "spidy",
               "Mercyful Fate",]
__license__ = 'ISC'
__version__ = '1.0rc1'
__maintainer__ = 'Jeff Quast'
__email__ = 'dingo@1984.ws'
__status__ = 'Development'

def main ():
    """
    x84 main entry point. The system begins and ends here.

    Command line arguments to engine.py:
      --config= location of alternate configuration file
      --logger= location of alternate logging.ini file
    """
    #pylint: disable=R0914
    #        Too many local variables (19/15)
    import getopt
    import sys

    import terminal
    import telnet
    import bbs.ini

    cfg_bbsfile = 'data/default.ini'
    cfg_logfile = 'data/logging.ini'
    try:
        opts, tail = getopt.getopt(sys.argv[1:], ":", ('config', 'logger',))
    except getopt.GetoptError, err:
        sys.stderr.write ('%s\n' % (err,))
        return 1
    for opt, arg in opts:
        if opt in ('--config',):
            cfg_bbsfile = arg
        elif opt in ('--logger',):
            cfg_logfile = arg
    assert 0 == len (tail), 'Unrecognized program arguments: %s' % (tail,)

    # load .ini files
    bbs.ini.init (cfg_bbsfile, cfg_logfile)

    # start telnet server
    addr_tup = (bbs.ini.CFG.get('telnet', 'addr'),
        int(bbs.ini.CFG.get('telnet', 'port')),)
    telnet_server = telnet.TelnetServer (
        address_pair = addr_tup,
        on_connect = terminal.on_connect,
        on_disconnect = terminal.on_disconnect,
        on_naws = terminal.on_naws)

    try:
        # begin main event loop
        _loop (telnet_server)
    except KeyboardInterrupt:
        for client in telnet_server.clients.values():
            client.deactivate ()
        telnet_server.poll ()
        raise SystemExit

def _loop(telnet_server):
    """
    Main event loop. Never returns.
    """
    # pylint: disable=R0912,R0914,R0915
    #         Too many branches (15/12)
    #         Too many local variables (24/15)
    #         Too many statements (73/50)
    import logging
    import time

    import terminal
    import db
    import bbs.ini

    logger = logging.getLogger()
    logger.info ('listening %s/tcp', telnet_server.port)
    client_timeout = int(bbs.ini.CFG.get('session', 'timeout', '1984'))
    locks = dict ()
    # main event loop
    while True:
        # process telnet i/o
        telnet_server.poll ()
        for client, pipe, lock in terminal.terminals():
            if not lock.acquire(False):
                continue
            lock.release ()

            # process telnet input (keypress sequences)
            if client.input_ready() and lock.acquire(False):
                lock.release()
                inp = client.get_input()
                pipe.send (('input', inp))

            # kick off idle users
            if client.idle() > client_timeout:
                logger.info ('%s timeout.', client.addrport())
                pipe.send (('exception', (
                    bbs.exception.ConnectionTimeout, None,)))
                client.deactivate ()
                continue

            if lock.acquire(False):
                # process bbs session i/o
                lock.release ()
                if not pipe.poll ():
                    continue

            # session i/o sent from child process
            try:
                event, data = pipe.recv()
            except EOFError:
                client.deactivate ()
                continue

            if event == 'disconnect':
                client.deactivate ()

            elif event == 'logger':
                logger.handle (data)

            elif event == 'output':
                text, cp437 = data
                if not cp437:
                    client.send_unicode (text)
                    continue
                else:
                    # disguise cp437 as ios8859-1 so bytes remain unmolested
                    bytestring = text.encode('iso8859-1', 'replace')
                    client.send_str (bytestring)

            elif event == 'global':
                #pylint: disable=W0612
                #         Unused variable 'o_lock'
                for o_client, o_pipe, o_lock in terminal.terminals():
                    if o_client != client:
                        o_pipe.send ((event, data,))

            #deprecation,
            #elif event == 'pos':
            #    assert type(data) in (float, int, type(None))
            #    # 'pos' query: 'what is the cursor position ?'
            #    # returns 'pos-reply' event as a callback
            #    # mechanism, data of (None, None) indicates timeout,
            #    # otherwise (y, x) is cursor position ..
            #    thread = terminal.POSHandler(pipe, client, lock,
            #        reply_event='pos-reply', timeout=data)
            #    thread.start ()

            elif event.startswith('db'):
                # db query-> database dictionary method, callback
                # with a matching ('db-*',) event. sqlite is used
                # for now and is quick, but this prevents slow
                # database queries from locking the i/o event loop.
                thread = db.DBHandler(pipe, event, data)
                thread.start ()

            elif event.startswith('lock'):
                # fine-grained lock acquire and release, non-blocking
                method, stale = data
                if method == 'acquire':
                    if not event in locks:
                        locks[event] = time.time ()
                        logger.debug ('(%r, %r) granted.', event, method)
                        pipe.send ((event, True,))
                    elif (stale is not None
                            and time.time() - locks[event] > stale):
                        logger.error ('(%r, %r) stale.', event, method)
                        pipe.send ((event, True,))
                    else:
                        logger.warn ('(%r, %r) failed.', event, method)
                        pipe.send ((event, False,))
                elif method == 'release':
                    if not event in locks:
                        logger.error ('(%s, %s) not acquired.', event, method)
                    else:
                        del locks[event]
                        logger.debug ('(%s, %s) removed.', event, data)

if __name__ == '__main__':
    exit(main ())
